experimental/ActionPolyRing/src/Operators.jl

```diff
@@ -0,0 +1,292 @@
+#
+#  Comparison 
+#
+###############################################################################
+
+### Difference ###
+Base.:(==)(dpre1::DifferencePolyRingElem{T}, dpre2::DifferencePolyRingElem{T}) where {T<:RingElement} = data(dpre1) == data(dpre2) && parent(dpre1) === parent(dpre2)
+
+function Base.hash(dpre::DifferencePolyRingElem, h::UInt)
+  b = 0x475b3fa701aa3148 % UInt
+  h = hash(parent(dpre), h)
+  h = hash(data(dpre), h)
+  return xor(h, b)
+end
+
+function Base.isless(dpre1::DifferencePolyRingElem{T}, dpre2::DifferencePolyRingElem{T}) where {T<:RingElement}
+  check_parent(dpre1, dpre2)
+  dpr = parent(dpre1)
+  vtj = __vtj(dpr)
+  @req haskey(vtj, dpre1) && haskey(vtj, dpre2) "Not jet variables in comparison"
+  m = nelementary_symbols(dpr)
+  ind1, ind2 = vtj[dpre1], vtj[dpre2]
+  v1, v2 = vcat([ind1[1] == j ? one(ZZ) : zero(ZZ) for j in 1:m], ind1[2]), vcat([ind2[1] == j ? one(ZZ) : zero(ZZ) for j in 1:m], ind2[2])
+  M = riquier_matrix(ranking(dpr))
+  return isless(M * v1, M * v2)
+end
+
+### Differential ###
+Base.:(==)(dpre1::DifferentialPolyRingElem{T}, dpre2::DifferentialPolyRingElem{T}) where {T<:RingElement} = data(dpre1) == data(dpre2) && parent(dpre1) === parent(dpre2)
+
+function Base.hash(dpre::DifferentialPolyRingElem, h::UInt)
+  b = 0x5c93cee72de560dd % UInt
+  h = hash(parent(dpre), h)
+  h = hash(data(dpre), h)
+  return xor(h, b)
+end
+
+function Base.isless(dpre1::DifferentialPolyRingElem{T}, dpre2::DifferentialPolyRingElem{T}) where {T<:RingElement}
+  check_parent(dpre1, dpre2)
+  dpr = parent(dpre1)
+  vtj = __vtj(dpr)
+  @req haskey(vtj, dpre1) && haskey(vtj, dpre2) "Not jet variables in comparison"
+  m = nelementary_symbols(dpr)
+  ind1, ind2 = vtj[dpre1], vtj[dpre2]
+  v1, v2 = vcat([ind1[1] == j ? one(ZZ) : zero(ZZ) for j in 1:m], ind1[2]), vcat([ind2[1] == j ? one(ZZ) : zero(ZZ) for j in 1:m], ind2[2])
+  M = riquier_matrix(ranking(dpr))
+  return isless(M * v1, M * v2)
+end
+
+###############################################################################
+#
+#  Unsafe functions 
+#
+###############################################################################
+
+### Difference ###
+function zero!(a::DifferencePolyRingElem{T}) where {T <: RingElement}
+  a.p = zero!(a.p)
+  return a
+end
+
+function one!(a::DifferencePolyRingElem{T}) where {T <: RingElement}
+  a.p = one!(a.p)
+  return a
+end
+
+function neg!(z::DifferencePolyRingElem{T}, a::DifferencePolyRingElem{T}) where {T <: RingElement}
+  if parent(data(z)) == parent(data(a))
+    z.p = neg!(z.p, a.p)
+  else
+    z.p = -a.p
+  end
+  return z
+end
+
+function fit!(a::DifferencePolyRingElem, n::Int)
+  fit!(data(a), n)
+end
+
+function add!(a::DifferencePolyRingElem{T}, b::DifferencePolyRingElem{T}, c::DifferencePolyRingElem{T}) where {T <: RingElement}
+  if parent(data(a)) == parent(data(b)) == parent(data(c))
+    a.p = add!(data(a), data(b), data(c))
+  else
+    a.p = data(b + c)
+  end
+  return a
+end
```

Comment:
you can even go one step further and merge Difference* and Differential* again, further reducing code duplication
```suggestion
function add!(a::PolyT, b::PolyT, c::PolyT) where {T <: RingElement, PolyT <: ActionPolyRingElem{T}}
  a.p = add!(data(a), data(b), data(c))
  return a
end
```